mpich: Update PMI configure options

raffenet opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
Add a "default" option that passes no option to configure. Existing
options changed in the MPICH 4.2.0 release, so update the package to
reflect those changes.

raffenet committed:
mpich: Fixup removal of pmi=off option

PR #42686 removed pmi=off but neglected to remove an instance of it in
the spec, and also broke a if/elif sequence.

raffenet committed:
mpich: Update PMI configure options

Add a "default" option that passes no option to configure. Existing
options changed in the MPICH 4.2.0 release, so update the package to
reflect those changes.

raffenet commented:
@spackbot fix style

raffenet commented:
@yfguo take a look when you have a chance. This is needed to add the new 4.2.0 release since the PMI options have changed.

alalazo reviewed (dismissed):
Feel free to ping me for a merge when ready

hzhou commented on var/spack/repos/builtin/packages/mpich/package.py:
```diff
@@ -61,9 +61,9 @@ class Mpich(AutotoolsPackage, CudaPackage, ROCmPackage):
+        default="default",
         description="""PMI interface.""",
-        values=("pmi", "pmi2", "pmix", "cray"),
+        values=("default", "pmi", "pmi2", "pmix", "cray"),
```
Add option `pmi1` for enable only `PMI-v1` protocol. I think `pmi=pmi` should be treated the same as `pmi=default`. In older mpich versions, `--with-pmi=simple` is the default anyway, right?
I think we also should add an option for "slurm", which should be configured as `"--with-pmilib=slurm --with-pmi2={0}".format{spec["slurm-pmi"].prefix`

raffenet replied:
I tried the suggested Slurm configuration among others but I just can't get it to work with the Slurm package. For some reason, it only installs `libpmi2.so`. I think we can probably modify the Slurm package to install both PMI1 and PMI2 and add MPICH support in a separate PR. That or we can default MPICH to use PMI2 with Slurm in `configure.ac`.

raffenet committed:
mpich: Rename pmi variant to pmi1

MPICH supports 3 PMI client interfaces - PMI1, PMI2, and PMIx. pmi=pmi
as a variant is poorly named, so make it more explicit.

raffenet commented:
Is this package audit failure caused by my patches?
```
Run . share/spack/setup-env.sh
  . share/spack/setup-env.sh
  $(which spack) audit packages
  $(which spack) audit externals
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.11.8/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.11.8/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.8/x64
    Python[2](https://github.com/spack/spack/actions/runs/8209487553/job/22455128621?pr=42885#step:6:2)_ROOT_DIR: /opt/hostedtoolcache/Python/[3](https://github.com/spack/spack/actions/runs/8209487553/job/22455128621?pr=42885#step:6:3).11.8/x6[4](https://github.com/spack/spack/actions/runs/8209487553/job/22455128621?pr=42885#step:6:4)
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.8/x[6](https://github.com/spack/spack/actions/runs/8209487553/job/22455128621?pr=42885#step:6:6)4
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.11.[8](https://github.com/spack/spack/actions/runs/8209487553/job/22455128621?pr=42885#step:6:8)/x64/lib
==> Error: cannot access local variable 'error_msg' where it is not associated with a value
Error: Process completed with exit code 3.
```

bernhardkaindl commented:
@yfguo and @hzhou, can you have a look if the new changes are fine for you now?

bernhardkaindl commented:
@raffenet 
> ==> Error: cannot access local variable 'error_msg' where it is not associated with a value

I just rebased this PR to the latest develop, so for some reason, this error appears from this PR.

Could you create a new PR and only add one small change at first to see if the error does not reappear?

If it does not reappear on the smallest change, you can progressively add more.

bernhardkaindl commented on var/spack/repos/builtin/packages/mpich/package.py:
```diff
@@ -551,16 +557,32 @@ def configure_args(self):
+                # use PMI2 interface with Cray PMI library
+                config_args.append(
+                    "--with-pmi=pmi2 --with-pmi2={0}".format(spec["cray-pmi"].prefix)
+                )
```
While it may work as you wrote it, just in case, it might be safer this way:
```suggestion
                config_args.append("--with-pmi=pmix")
                config_args.append(f"--with-pmix={spec['pmix'].prefix}")
            elif "pmi=cray" in spec:
                # use PMI2 interface of the Cray PMI library
                config_args.append("--with-pmi=pmi2")
                config_args.append(f"--with-pmi2={spec['cray-pmi'].prefix}")
```
PS: Yes, you should use f-strings in `spack` now.

bernhardkaindl commented on var/spack/repos/builtin/packages/mpich/package.py:
```diff
@@ -551,16 +557,32 @@ def configure_args(self):
+            elif "pmi=pmi2" in spec:
+                config_args.append("--with-pmi=pmi2/simple")
+            elif "pmi=pmix" in spec:
+                config_args.append("--with-pmix={0}".format(spec["pmix"].prefix))
```
Likewise, use an f-string:
```suggestion
                config_args.append(f"--with-pmix={spec['pmix'].prefix}")
```

bernhardkaindl reviewed (commented):
Use f-strings and split the appends for the options (just in case, and nicer)

bernhardkaindl committed:
Apply suggestions from code review: Split appends an use f-strings

bernhardkaindl reviewed (commented):
Applied the use of f-strings and spitting of the appends.

@raffenet: Converted to draft, proposing to start a new PR, initially with only the smallest change possible.
@raffenet: Please close this PR when all changes have been moved to one or more new PRs (if they don't have this setup issue)

hzhou reviewed (approved):
LGTM
